### PR TITLE
v4.2.0 - Update on save

### PR DESCRIPTION
**Pull request by ivan-doofinder**
- Add configuration option in backend to configure update on save options
- Created hooks to collect modified products
- Saving modified products in queue
- Create builder to generate payload to send to the API
- Get search engines when saving update on save configuration
- Process product queue
- Created wrapper for sending products to the API
- Upgrade module file
- Fixes product builder 
- Fix PHP formatting

### CODE REVIEW

File: doofinder.php [hunk 1]
```diff
@@ -829,6 +857,28 @@ protected function getConfigFormDataFeed()
...
+                    array(
+                        'type' => 'select',
+                        'label' => $this->l('Update on save delay'),
+                        'name' => 'DF_UPDATE_ON_SAVE_DELAY',
+                        'options' => array(
+                            'query' => array(
+                                90 => array("id" => 90, "name" => "90 minutos"),
+                                60 => array("id" => 60, "name" => "60 minutos"),
+                                30 => array("id" => 30, "name" => "30 minutos"),
+                                15 => array("id" => 15, "name" => "15 minutos")
```
Comment (alvanieto) [inline doofinder.php]: minutos other minutes?

File: doofinder.php [hunk 2]
```diff
@@ -1690,6 +1747,149 @@ public function hookProductSearchProvider($params)
...
+        );
+
+        return array_column($products, "id_product");
+    }
+
+    public function deleteProductQueue($id_shop){
+        Db::getInstance()->execute("DELETE from " . _DB_PREFIX_ . "doofinder_product WHERE id_shop = " . (int)$id_shop);
+    }
+
+    public function sendProductsApi($products, $id_shop, $id_lang, $id_currency, $action = "update")
```
Comment (alvanieto) [inline doofinder.php:1842]: Two explicit functions, one for update and the other for delete and a third functions with the common code to be called in update & delete?

Comment (ivan-doofinder) [reply doofinder.php:1842]: The **hookActionProductSave** and **hookActionProductDelete** functions are the ones triggered by Prestashop when the corresponding modification is made to the product. This is managed by Prestashop like this.

Then a function is called to add to the queue, which is common for update/remove.
And then a function is called to process the queue, which processes both cases (modified and deleted).

File: doofinder.php (hunk 1 above)
Comment (ivan-doofinder) [reply doofinder.php]: I have to add translations and texts. I am waiting for Cristobal to confirm the texts to be displayed for each language.

Comment (ivan-doofinder) [reply doofinder.php]: translated!

File: lib/dfProduct_build.php [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+            $p = array_merge($p, $attributes);
+        }
+
+        return $p;
+    }
+
+    private function getId($product)
+    {
+        if ($this->product_variations) {
+            if (isset($product['id_product_attribute']) && (int)$product['id_product_attribute'] > 0) {
```
Comment (abraham-doofinder) [inline lib/dfProduct_build.php]: this code is repeated in various functions, maybe we should decouple in another function

Comment (ivan-doofinder) [reply lib/dfProduct_build.php]: replaced by a function to avoid repeating code

**Pull request merged**